components/motor/ulnstepper/28byj-48.go

```diff
@@ -0,0 +1,478 @@
+	In4 string `json:"In4"`
+}
+
+// Config describes the configuration of a motor.
+type Config struct {
+	Pins              PinConfig `json:"pins"`
+	BoardName         string    `json:"board"`
+	RotationPerMinute float64   `json:"rotation_per_minute"`
+	TicksPerRotation  int       `json:"ticks_per_rotation"`
+}
+
+// Validate ensures all parts of the config are valid.
+func (config *Config) Validate(path string) ([]string, error) {
+	var deps []string
+	if config.BoardName == "" {
+		return nil, utils.NewConfigValidationFieldRequiredError(path, "board")
+	}
+	deps = append(deps, config.BoardName)
+	return deps, nil
+}
+
+func init() {
+	_motor := registry.Component{
+		Constructor: func(ctx context.Context, deps registry.Dependencies, config config.Component, logger golog.Logger) (interface{}, error) {
+			actualBoard, motorConfig, err := getBoardFromRobotConfig(deps, config)
+			if err != nil {
+				return nil, err
+			}
+
+			return newULN(ctx, actualBoard, *motorConfig, config.Name, logger)
+		},
+	}
+	registry.RegisterComponent(motor.Subtype, model, _motor)
+	config.RegisterComponentAttributeMapConverter(
+		motor.Subtype,
+		model,
+		func(attributes config.AttributeMap) (interface{}, error) {
+			var conf Config
+			return config.TransformAttributeMapToStruct(&conf, attributes)
+		},
+		&Config{},
+	)
+}
+
+func getBoardFromRobotConfig(deps registry.Dependencies, config config.Component) (board.Board, *Config, error) {
+	motorConfig, ok := config.ConvertedAttributes.(*Config)
+	if !ok {
+		return nil, nil, rdkutils.NewUnexpectedTypeError(motorConfig, config.ConvertedAttributes)
+	}
+	if motorConfig.BoardName == "" {
+		return nil, nil, errors.New("expected board name in config for motor")
+	}
+	b, err := board.FromDependencies(deps, motorConfig.BoardName)
+	if err != nil {
+		return nil, nil, err
+	}
+	return b, motorConfig, nil
+}
+
+func newULN(ctx context.Context, b board.Board, mc Config, name string, logger golog.Logger) (motor.Motor, error) {
+	if mc.TicksPerRotation == 0 {
+		return nil, errors.New("expected ticks_per_rotation in config for motor")
+	}
+
+	if mc.RotationPerMinute == 0 {
+		return nil, errors.New("expected rotation_per_minute value greater than zero")
+	}
+
+	m := &uln2003{
+		theBoard:          b,
+		rotationPerMinute: mc.RotationPerMinute,
+		ticksPerRotation:  mc.TicksPerRotation,
+		logger:            logger,
+		motorName:         name,
+	}
+
+	if mc.Pins.In1 != "" {
+		in1, err := b.GPIOPinByName(mc.Pins.In1)
+		if err != nil {
+			return nil, err
+		}
+		m.in1 = in1
+	}
+
+	if mc.Pins.In2 != "" {
+		in2, err := b.GPIOPinByName(mc.Pins.In2)
+		if err != nil {
+			return nil, err
+		}
+		m.in2 = in2
+	}
+
+	if mc.Pins.In3 != "" {
+		in3, err := b.GPIOPinByName(mc.Pins.In3)
+		if err != nil {
+			return nil, err
+		}
+		m.in3 = in3
+	}
+
+	if mc.Pins.In4 != "" {
+		in4, err := b.GPIOPinByName(mc.Pins.In4)
+		if err != nil {
+			return nil, err
+		}
+		m.in4 = in4
+	}
+
+	if err := m.Validate(); err != nil {
```

Comment:
[minor] `Validate` should be called on config rather than in this function. Follows our more updated rules. I realise `gpiostepper` is different, but I'd like this structural change.